aead/src/lib.rs

```diff
@@ -66,64 +60,48 @@ pub trait Aead {
 /// scheme. This is the standard RFC algorithm.
 pub trait StatelessAead {
     /// The key size in a new method.
-    type KeyLength: ArrayLength<u8> + Unsigned;
-    /// The maximum length a plaintext can be.
-    type PlaintextMax: ArrayLength<u8> + Unsigned;
-    /// The maximum length a ciphertext can be.
-    type CiphertextMax: ArrayLength<u8> + Unsigned;
-    /// The maximum length of the associated data.
-    type AssociatedDataMax: ArrayLength<u8> + Unsigned;
-    /// The minimum length of the nonce.
-    type NonceMin: ArrayLength<u8> + Unsigned;
+    type KeySize: ArrayLength<u8> + Unsigned;
+    /// The length of a nonce.
+    type NonceSize: ArrayLength<u8> + Unsigned;
     /// The maximum length of the nonce.
-    type NonceMax: ArrayLength<u8> + Unsigned;
-
-    /// An actual type indicating the nonce
-    type Nonce;
-
-    /// Retrieve the size of the buffer required for plaintext of a given size.
-    fn ciphertext_len(plaintext_used: usize) -> usize;
+    type TagSize: ArrayLength<u8> + Unsigned;
+    /// The amount of suffix padding, in bytes, which needs to be
+    /// appended to a plaintext to accomodate this cipher's output.
+    type CiphertextOverhead: ArrayLength<u8> + Unsigned;
 
     /// Encrypts the given plaintext into a new ciphertext object and the nonce
     fn encrypt<'in_out, AdItem: AsRef<[u8]>, AdIter: Iterator<Item = AdItem>>(
-        key: &GenericArray<u8, Self::KeyLength>,
+        key: &GenericArray<u8, Self::KeySize>,
         additional_data: AdIter,
```

Comment:
I'd suggest using `impl AsRef<[u8]>` here, rather than generic parameters or an iterator.

I can't think of any constructions that take a vector of associated data besides S2V-based constructions like AES-SIV, and regardless, I think it's needless complexity in a standard AEAD API.

Reply:
I like this because it is very useful for building encapsulated protocols, where I am able to authenticate lower-level protocol details using the application-level channel without a lot of copying and fussing of those buffers.

Reply:
I have never seen another AEAD API which acts on a vector of AD, and can't think of any AEAD algorithms that support a vector of AEAD besides S2V-based constructions.

What are all of the normal AEAD modes that take only a single bytestring as AD supposed to do here when given more than one bytestring? Return an error?

If the goal is to avoid concatenating it into a contiguous slice, I'd argue that needs an API which treats all parameters that way, which is much more complex than what we're trying to spec here. KISS.

Reply:
Blech, OK, I wasn't aware other modes could only ever take a single string.